source/PluginDev/Assets/GooglePlayGames/Platforms/Android/AndroidRealTimeMultiplayerClient.cs

```diff
@@ -34,111 +56,689 @@ public AndroidRealTimeMultiplayerClient(AndroidJavaObject account)
+                    /* callback= */ null))
+                    ;
+                return;
+            }
+
+            using (mRtmpClient.Call<AndroidJavaObject>("sendUnreliableMessage", data, roomId, participantId)) ;
         }
 
         public void SendMessage(bool reliable, string participantId, byte[] data, int offset, int length)
         {
-            // Task<Integer> sendReliableMessage(@NonNull byte[] messageData, @NonNull String roomId, @NonNull String recipientParticipantId, @Nullable ReliableMessageSentCallback callback)
-            // Task<Void> sendUnreliableMessage(@NonNull byte[] messageData, @NonNull String roomId, @NonNull String recipientParticipantId)
+            SendMessage(reliable, participantId, Misc.GetSubsetBytes(data, offset, length));
         }
 
         public List<Participant> GetConnectedParticipants()
         {
-            return null;
+            List<Participant> result = new List<Participant>();
+            foreach (Participant participant in GetParticipantList())
+            {
+                if (participant.IsConnectedToRoom)
+                {
+                    result.Add(participant);
+                }
+            }
+
+            return result;
+        }
+
+        private List<Participant> GetParticipantList()
+        {
+            if (mRoom == null)
+            {
+                return new List<Participant>();
+            }
+
+            List<Participant> participants = new List<Participant>();
+            using (var participantsObject = mRoom.Call<AndroidJavaObject>("getParticipants"))
+            {
+                int size = participantsObject.Call<int>("size");
+                for (int i = 0; i < size; i++)
+                {
+                    using (var participant = participantsObject.Call<AndroidJavaObject>("get", i))
+                    {
+                        participants.Add(AndroidJavaConverter.ToParticipant(participant));
+                    }
+                }
+            }
+
+            return participants;
         }
 
         public Participant GetSelf()
         {
+            if (GetRoomStatus() != RoomStatus.Active)
+            {
+                return null;
+            }
+
+            foreach (var participant in GetParticipantList())
+            {
+                if (participant.Player.id.Equals(mAndroidClient.GetUserId()))
+                {
+                    Debug.Log("self id is " + participant.ParticipantId);
+                    return participant;
+                }
+            }
+
             return null;
         }
 
         public Participant GetParticipant(string participantId)
         {
-            return null;
+            if (GetRoomStatus() != RoomStatus.Active)
+            {
+                return null;
+            }
+
+            using (var participant = mRoom.Call<AndroidJavaObject>("getParticipant", participantId))
```

Comment:
I suspect getParticipant can return null in this case we need try catch block

Reply:
It doesn't return null but it throws exception when not found. But it's better for us to have it null here.